stub/src/main/java/io/grpc/stub/BlockingClientCall.java

```diff
@@ -0,0 +1,407 @@
+import java.util.logging.Level;
+import java.util.logging.Logger;
+
+/**
+ * Represents a bidirectional streaming call from a client.  Allows in a blocking manner, sending
+ * over the stream and receiving from the stream.  Also supports terminating the call.
+ * Wraps a ClientCall and converts from async communication to the public sync paradigm.
+ *
+ * @param <ReqT> Type of the Request Message
+ * @param <RespT> Type of the Response Message
+ */
+public final class BlockingClientCall<ReqT, RespT> {
+
+  private static final Logger logger = Logger.getLogger(BlockingClientCall.class.getName());
+
+  private final BlockingQueue<RespT> buffer;
+  private final StartableListener<RespT> listener;
+  private final ClientCall<ReqT, RespT> call;
+
+  private final ThreadlessExecutor executor;
+
+  private volatile boolean writeClosed;
+  private Status closedStatus;
+
+  BlockingClientCall(ClientCall<ReqT, RespT> call, ThreadlessExecutor executor) {
+    this.call = call;
+    this.executor = executor;
+    buffer = new ArrayBlockingQueue<>(1);
+    listener = new QueuingListener(buffer, call);
+  }
+
+  /**
+   * Wait if necessary for a value to be available from the server. If there is an available value
+   * return it immediately, if the stream is closed return a null. Otherwise, wait for a value to be
+   * available or the stream to be closed
+   *
+   * @return value from server or null if stream has been closed
+   * @throws io.grpc.StatusRuntimeException If the stream has closed in an error state
+   */
+  public RespT read() throws InterruptedException {
+    try {
+      return read(Integer.MAX_VALUE, TimeUnit.DAYS);
+    } catch (TimeoutException e) {
+      throw new RuntimeException(e); // should never happen
+    }
+  }
+
+  /**
+   * Wait with timeout, if necessary, for a value to be available from the server. If there is an
+   * available value, return it immediately.  If the stream is closed return a null. Otherwise, wait
+   * for a value to be available, the stream to be closed or the timeout to expire.
+   *
+   * @param timeout how long to wait before giving up.  Values &lt;= 0 are no wait
+   * @param unit a TimeUnit determining how to interpret the timeout parameter
+   * @return value from server or null (if stream has been closed)
+   * @throws TimeoutException if no read becomes ready before the specified timeout expires
+   * @throws io.grpc.StatusRuntimeException If the stream has closed in an error state
+   */
+  public RespT read(long timeout, TimeUnit unit) throws InterruptedException, TimeoutException {
+    long start = System.nanoTime();
+
+    executor.drain();
+
+    if (buffer.isEmpty() && closedStatus != null) {
+      if (closedStatus.isOk()) {
+        return null;
+      } else {
+        throw closedStatus.asRuntimeException();
+      }
+    }
+
+    long duration = unit.toNanos(timeout);
+    RespT bufferedValue;
+    while ((bufferedValue = buffer.poll()) == null && closedStatus == null) {
```

Comment:
I don't think this pattern works for multiple consuming threads. Seems there is a race between the poll and the blocking.

transport | read() | write()
--------- | ------ | -------
||executor.drain()
executor.execute(onMessage)
||buffer.isEmpty() == true
||buffer.poll() == null
|||call.isReady() == false
|||waitAndDrainExecutorOrTimeout()
|||onMessage() {buffer.add(msg)}
||waitAndDrainExecutorOrTimeout()
||frozen, but buffer contains entry

Reply:
Introduced a predicate with target object that allows for validation inside synchronized block so that waitAndDrainExecutorOrTimeout returns without blocking.  To deal with the onMessage arriving in the middle of waitAndDrainExecutorOrTimeout and being processed outside the synchronized block by the do-while, there is an extra pass after that in a synchronized block of the threads with predicates to wake them up if ready.